File: 0567-permutation-in-string/0567-permutation-in-string.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        int matches = 0;
        
        if(s1.size() > s2.size()) return false;
        
        vector<int> counter1(26, 0);
        vector<int> counter2(26, 0);
        
        for(int i=0; i<s1.size(); i++){
            counter1[s1[i]-'a']++;
            counter2[s2[i]-'a']++;
        }
        
        for(int i=0; i<26; i++){
            if(counter1[i] == counter2[i]) matches++;
        }
        
        int li=0;
        int ri=s1.size();
        
        while(ri < s2.size()){
            if(matches == 26) return true;
            
            counter2[s2[ri]-'a']++;
            
            
            if(counter1[s2[ri]-'a'] == counter2[s2[ri]-'a']) matches++;
            else if(counter1[s2[ri]-'a']+1 == counter2[s2[ri]-'a']) matches--;
            
            counter2[s2[li]-'a']--;
            
            if(counter1[s2[li]-'a'] == counter2[s2[li]-'a']) matches++;
            else if(counter1[s2[li]-'a']-1 == counter2[s2[li]-'a'])matches--;
            
            ri++;
            li++;
        }
        
        return matches==26;
    }
};

File: 0567-permutation-in-string/0567-permutation-in-string.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        if len(s1) > len(s2):
            return False
        
        freqS1 = [0] * 26
        
        for s in s1:
            freqS1[ord(s) - ord('a')] += 1

        window_size = len(s1)
        window = [0] * 26
        
        for right in range(window_size):
            window[ord(s2[right]) - ord('a')] += 1
        
        if window == freqS1: return True
        
        left = 0
        
        for right in range(window_size, len(s2)):
            window[ord(s2[right]) - ord('a')] += 1
            window[ord(s2[left]) - ord('a')] -= 1
            
            left += 1

            if window == freqS1: return True

        return False
        